### PR TITLE
Remove premature optimization to make request source factory more reusable

### PR DESCRIPTION
**Pull request by wjuan-AFK**
Fixing issue where nighthawk makes request source hard to reuse due to caching the options_list and removing associated todo. Caching a single list in the factory was a premature optimization and resulted in nighthawk being hard to reuse. Also changing so that an empty optionslist will return the default header instead of throwing an error. Added associated test for loading two request sources with the same factory.

### CODE REVIEW

**Comment by wjuan-AFK** [comment]
Ran test_gcc manually and passed.

File: test/request_source/request_source_plugin_test.cc [hunk 1]
```diff
@@ -337,5 +394,75 @@ TEST_F(InLineRequestSourcePluginTest,
...
+
+  Envoy::ProtobufWkt::Any config_ab_any;
+  config_ab_any.PackFrom(config_ab);
+  Envoy::ProtobufWkt::Any config_c_any;
+  config_c_any.PackFrom(config_c);
+
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+          "nighthawk.in-line-options-list-request-source-plugin");
+  auto header_ab = Envoy::Http::RequestHeaderMapImpl::create();
```
Comment (jiajunye) [inline test/request_source/request_source_plugin_test.cc:430]: Can we use the actual std::unique_ptr<TerminationPredicate> type here and below?

File: test/request_source/request_source_plugin_test.cc [hunk 2]
```diff
@@ -223,6 +223,63 @@ TEST_F(FileBasedRequestSourcePluginTest,
...
+
+  Envoy::ProtobufWkt::Any config_ab_any;
+  config_ab_any.PackFrom(config_ab);
+  Envoy::ProtobufWkt::Any config_c_any;
+  config_c_any.PackFrom(config_c);
+
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<RequestSourcePluginConfigFactory>(
+          "nighthawk.file-based-request-source-plugin");
+  auto header_ab = Envoy::Http::RequestHeaderMapImpl::create();
```
Comment (jiajunye) [inline test/request_source/request_source_plugin_test.cc:246]: Can we use the actual std::unique_ptr<TerminationPredicate> type here and below?

**Review by jiajunye — APPROVED** [review]
LGTM

**Review by mum4k — APPROVED** [review]

**Comment by dubious90** [comment]
Once Jason's two items are addressed we can merge this.  @wjuan-AFK please address and assign to me for merge (mum4k is out)

**Review by dubious90 — APPROVED** [review]

**Pull request merged**